### PR TITLE
Constant parameter - Fix #117

### LINKED ISSUE #117: Bug when using a constant for a scalar kernel parameter
When a kernel is invoked as:
```
  call invoke( bc_ssh(0, p_fld) )
```
with a constant value as the scalar parameter  (bc_ssh is unchanged from the test files):
```
  type, extends(kernel_type) :: bc_ssh
     type(arg), dimension(3) :: meta_args =        &
          (/ arg(READ,      R_SCALAR, POINTWISE),  &
             arg(READWRITE, CT,       POINTWISE),  &
             arg(READ,      GRID_MASK_T)           &
           /)
```
the code created includes:
```
REAL(KIND=wp), intent(inout) :: 0
```
In https://github.com/hiker/PSyclone/tree/constant-parameter I've created an x-failing test to show this, not sure if you want a pull request for the test itself. Easy work around of course: declare a variable zero=0.

### PR DESCRIPTION
**Pull request by hiker**
Well, that's all to say: fix constant parametes in gocean, see #117 for details.

### CODE REVIEW

File: src/psyclone/gocean1p0.py [hunk 1]
```diff
@@ -169,7 +169,8 @@ def unique_args_rscalars(self):
         for call in self._schedule.calls():
             for arg in call.arguments.args:
                 if arg.type == 'scalar' and \
-                   arg.space.lower() == "r_scalar" and arg.name not in result:
+                   arg.space.lower() == "r_scalar" and \
```
Comment (arporter) [inline src/psyclone/gocean1p0.py]: Changes to exclude literal constants - good.

File: src/psyclone/tests/gocean1p0_test.py [hunk 2]
```diff
@@ -972,6 +974,20 @@ def test00p1_kernel_wrong_meta_arg_count():
               api="gocean1.0")
 
 
+def test00p1_invoke_kernel_using_const_scalar():  # pylint:disable=invalid-name
+    '''Check' that using a const scalar as parameter works.'''
```
Comment (arporter) [inline src/psyclone/tests/gocean1p0_test.py]: Superfluous "'" after "Check"

File: src/psyclone/tests/test_files/gocean1p0/test00.1_invoke_kernel_using_const_scalar.f90 [hunk 3]
```diff
@@ -0,0 +1,24 @@
+!-------------------------------------------------------------------------------
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/test00.1_invoke_kernel_using_const_scalar.f90]: Please change this copyright notice to the one used now (e.g. in `test_files/dynamo0p3/testkern.F90`) and update the date to 2018.

File: src/psyclone/tests/test_files/gocean1p0/test00.1_invoke_kernel_using_const_scalar.f90 [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+! (c) The copyright relating to this work is owned jointly by the Crown,
+! Met Office and NERC 2015.
+! However, it has been created with the help of the GungHo Consortium,
+! whose members are identified at https://puma.nerc.ac.uk/trac/GungHo/wiki
+!-------------------------------------------------------------------------------
+! Author A. Porter STFC Daresbury Lab
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/test00.1_invoke_kernel_using_const_scalar.f90]: Please make yourself the author and remove the bit about funding from the GOcean project.

File: src/psyclone/tests/gocean1p0_test.py [hunk 5]
```diff
@@ -972,6 +974,20 @@ def test00p1_kernel_wrong_meta_arg_count():
...
+                            "test00.1_invoke_kernel_using_const_scalar.f90")
+    _, invoke_info = parse(filename, api="gocean1.0")
+    out = str(PSyFactory(API).create(invoke_info).gen)
+    # Old versions of PSyclone tried to declare '0' as a variable:
+    # REAL(KIND=wp), intent(inout) :: 0
+    # Make sure this is not happening anymor
+    import re
+    assert re.search(r"\s*real.*:: *0", out, re.I) is None
+
+
```
Comment (arporter) [inline src/psyclone/tests/gocean1p0_test.py]: Please add a test for passing a real quantity by value. There may already be some test Fortran code where this is done.

Comment (hiker) [reply src/psyclone/tests/gocean1p0_test.py]: OK, I am sure I misunderstood something ... I've added a test with a kernel using a by-value parameters, but I am somehow that doesn't make any sense tbh, as far as I can see this has nothing to do with this pull request. I've added this as a separate commit, so it's easy to undo. Just let me know what I missed here ;)

**Comment by hiker** [comment]
Besides my comment on adding a by-value test above, everything else should be done.

File: src/psyclone/tests/gocean1p0_test.py [hunk 6]
```diff
@@ -973,6 +975,21 @@ def test00p1_kernel_wrong_meta_arg_count():
...
+                            "test_files", "gocean1p0",
+                            "test00.1_invoke_kernel_using_const_scalar.f90")
+    _, invoke_info = parse(filename, api="gocean1.0")
+    out = str(PSyFactory(API).create(invoke_info).gen)
+    # Old versions of PSyclone tried to declare '0' as a variable:
+    # REAL(KIND=wp), intent(inout) :: 0
+    # Make sure this is not happening anymor
+    import re
+    assert re.search(r"\s*real.*:: *0", out, re.I) is None
+    assert re.search(r"\s*real.*:: *real_val", out, re.I) is not None
```
Comment (arporter) [inline src/psyclone/tests/gocean1p0_test.py]: :-) When I said 'real' I meant try testing cases where both a real quantity e.g. "0.0_dp" and an integer quantity, e.g. "1" are passed by value. I realise that, on looking at this again the thing being passed as "0" is in fact a real variable so my request was confusing. Does it make sense now?

**Comment by arporter** [comment]
Almost there. Changes are pep8 and pylint clean. Just need to extend test to explicitly test passing both integer and floating-point quantities by value.

File: src/psyclone/tests/gocean1p0_test.py (hunk 6 above)
Comment (hiker) [reply src/psyclone/tests/gocean1p0_test.py]: *grin* the 'by value' caused me confusion. I added an integer parameter to the new test function, and passed both a constant 0 as real and int in. I left the test with by-value in place, can't really hurt as far as I can tell.

**Comment by arporter** [comment]
All comments have been addressed. Tests all OK and new code is covered. All examples run OK.

**Pull request merged**